rust/src/http2/parser.rs

```diff
@@ -0,0 +1,1483 @@
+    } else {
+        match http2_frame_header_static(indexed.1) {
+            Some(x) => Ok((i2, x.name)),
+            None => Err(Err::Error((i2, ErrorKind::MapOpt))),
+        }
+    }?;
+    let (i4, value) = http2_parse_headers_block_string(i3)?;
+    return Ok((i4, HTTP2FrameHeaderBlock { name, value }));
+}
+
+named!(
+    http2_parse_headers_block<HTTP2FrameHeaderBlock>,
+    alt!(
+        http2_parse_headers_block_indexed
+            | http2_parse_headers_block_literal_incindex
+            | http2_parse_headers_block_literal_noindex
+            | http2_parse_headers_block_literal_neverindex
+    )
+);
+//TODO 6.3 Dynamic Table Size Update
+
+#[derive(Clone)]
+pub struct HTTP2FrameHeaders {
+    pub padlength: Option<u8>,
+    pub priority: Option<HTTP2FrameHeadersPriority>,
+    pub blocks: Vec<HTTP2FrameHeaderBlock>,
+}
+
+const HTTP2_FLAG_HEADER_PADDED: u8 = 0x8;
+const HTTP2_FLAG_HEADER_PRIORITY: u8 = 0x20;
+
+pub fn http2_parse_frame_headers(input: &[u8], flags: u8) -> IResult<&[u8], HTTP2FrameHeaders> {
+    do_parse!(
+        input,
+        padlength: cond!(flags & HTTP2_FLAG_HEADER_PADDED != 0, be_u8)
+            >> priority:
+                cond!(
+                    flags & HTTP2_FLAG_HEADER_PRIORITY != 0,
+                    http2_parse_headers_priority
+                )
+            >> blocks: many0!(http2_parse_headers_block)
+            >> (HTTP2FrameHeaders {
+                padlength,
+                priority: priority,
+                blocks: blocks
+            })
+    )
+}
+//TODO keep reading headers even if one has issues
+
+#[repr(u16)]
+#[derive(Clone, Copy, PartialEq, FromPrimitive, Debug)]
+pub enum HTTP2SettingsId {
+    SETTINGSHEADERTABLESIZE = 1,
+    SETTINGSENABLEPUSH = 2,
+    SETTINGSMAXCONCURRENTSTREAMS = 3,
+    SETTINGSINITIALWINDOWSIZE = 4,
+    SETTINGSMAXFRAMESIZE = 5,
+    SETTINGSMAXHEADERLISTSIZE = 6,
+}
+
+impl fmt::Display for HTTP2SettingsId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::str::FromStr for HTTP2SettingsId {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let su = s.to_uppercase();
+        let su_slice: &str = &*su;
+        match su_slice {
+            "SETTINGS_HEADER_TABLE_SIZE" => Ok(HTTP2SettingsId::SETTINGSHEADERTABLESIZE),
+            "SETTINGS_ENABLE_PUSH" => Ok(HTTP2SettingsId::SETTINGSENABLEPUSH),
+            "SETTINGS_MAX_CONCURRENT_STREAMS" => Ok(HTTP2SettingsId::SETTINGSMAXCONCURRENTSTREAMS),
+            "SETTINGS_INITIAL_WINDOW_SIZE" => Ok(HTTP2SettingsId::SETTINGSINITIALWINDOWSIZE),
+            "SETTINGS_MAX_FRAME_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXFRAMESIZE),
+            "SETTINGS_MAX_HEADER_LIST_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXHEADERLISTSIZE),
+            _ => Err(format!("'{}' is not a valid value for HTTP2SettingsId", s)),
+        }
+    }
+}
+
+//TODO move elsewhere generic
+#[derive(PartialEq, Debug)]
```

Comment:
Thoughts about this ?

Reply:
About what specifically?

Reply:
Should I move this code elsewhere ?
And if it is the case, where precisely ?

This code is basically the Rust translation of detect-engine-uint.c